### ISSUE TITLE
Issues with remove deleted

### ISSUE THREAD

**Issue by Ramon-zaro** [dadoonet/fscrawler#531]
Hi David,
I've been having problems with this option, in that it is not removing from index references to files already deleted and/or moved from the indexed drive. In many cases the parent folder itself does not exist anymore.
I have tried restarting but no luck so far.

**Comment by dadoonet** [comment]
For sure `restart` option in the case of deleted documents will not have any effect. I'm using the date stored in the status file to check what changed after last run including deletes.

**Comment by Ramon-zaro** [comment]
okay, so how is the option supposed to work ?

**Comment by dadoonet** [comment]
It is supposed to detect that the directory which contains files or dirs that have been removed has an updated date.
In which case we check against elasticsearch content.

Note that if you have more than 10000 documents removed that might not work well for now.
I expect some optimizations in the future as I want to change the implementation of this with #399

**Comment by Ramon-zaro** [comment]
Hi David, 
moving on from the discussion we had here : https://discuss.elastic.co/t/max-content-length-issue/125016/3

I was under the impression that I needed to set store_source to "true" for the remove deleted option to work. is that correct ?

Let me rephrase my requirements; I want to : 
a) index complete file content, not just 100,000 characters
AND
b) drop moved/deleted files & folders from index.

My current deployment has to cater to about a million or so files, mostly of the doc, pdf etc type.

What are my best options to achieve the above ? Do I need to move to FS 2.5 ?

**Comment by dadoonet** [comment]
Hey @Ramon-zaro. Sorry for the delay. Has been kind of crazy year (as usual :) )

> I was under the impression that I needed to set store_source to "true" for the remove deleted option to work. is that correct ?

No. I don't think so as only metadata are needed to make that work and not the binary document itself.

> index complete file content, not just 100,000 characters 

In that case you must set `indexed_chars` to `-1`. See https://github.com/dadoonet/fscrawler#extracted-characters

> drop moved/deleted files & folders from index

That should work out of the box but may be there are some issues with some file systems. That's why I need to change the implementation at some point.

**Comment by Ramon-zaro** [comment]
thanks a lot. you are providing a great solution to a pressing real life problem, so I can't really complain. 
I will run some tests and get back to you.

1 thing I do hope is there in 2.5 is an option to divide the indexing by size rather than by number of files, as in the current implementation. because of wide variation in file sizes even if an insignificant number of files are oversized it can choke up either FS or ES, so the only option is to se bulk size to '1', which makes for slow indexing.

**Comment by dadoonet** [comment]
Can you open an issue to specifically add support for bulk size in mb or kb?

**Comment by dadoonet** [comment]
> 1 thing I do hope is there in 2.5 is an option to divide the indexing by size rather than by number of files, as in the current implementation. because of wide variation in file sizes even if an insignificant number of files are oversized it can choke up either FS or ES, so the only option is to se bulk size to '1', which makes for slow indexing.

Done with #563. I wonder if this issue can be closed now.

**Comment by Ramon-zaro** [comment]
excellent ! I will try it out and report.

**Comment by Ramon-zaro** [comment]
really sorry for the enormously late response. the option works beatifully. 

there's still a problem with remove deleted though. (as in it is not working as advertised).
also, if I try restarting with remove_deleted set to true after initially indexing with it set to true, it throws an error message. 

I wonder if there is a way of removing deleted files from an index that has initially created an index where it was not set to true.

**Comment by dadoonet** [comment]
It should work IMO but this is a scenario I'm not testing. I need to add a test probably.

Can you run in debug mode so I can check how all this is working?

Note that if you upgrade at the same time from an old version to a new one that might explain. If so, please tell me.

**Comment by Ramon-zaro** [comment]
you hint worked, it was indeed a problem with upgrade. it was missing the filename stored mapping.

**Comment by dadoonet** [comment]
I believe I can close this one.
